### PR TITLE
MRG: Faster basis calculation

### PR DESCRIPTION
**Pull request by Eric89GXL**
Some minor tweaks to make Maxwell filtering a tiny bit faster. Also another validation of our basis calculations.

@wronk you want to have a look? Thanks @jnenonen and @staulu for the discussions about all this.

### CODE REVIEW

**Comment by agramfort** [comment]
nobody is really able to give you feed back here. 

if you say that results are the same and computation time is lower please merge.

File: mne/preprocessing/maxwell.py [hunk 1]
```diff
@@ -678,6 +650,209 @@ def _sss_basis(origin, coils, int_order, ext_order, mag_scale=100.,
...
+    coils : list
+        List of MEG coils. Each should contain coil information dict specifying
+        position, normals, weights, number of integration points and channel
+        type. All coil geometry must be in the same coordinate frame
+        as ``origin`` (``head`` or ``meg``).
+    int_order : int
+        Order of the internal multipolar moment space
+    ext_order : int
+        Order of the external multipolar moment space
+    scale : float
```
Comment (jasmainak) [inline mne/preprocessing/maxwell.py]: what is this documenting? I don't see this as an argument. How are the tests passing if this is the case?

File: mne/preprocessing/maxwell.py [hunk 2]
```diff
@@ -678,6 +650,209 @@ def _sss_basis(origin, coils, int_order, ext_order, mag_scale=100.,
...
+        Scale factor for magnetometers.
+    method : str
```
Comment (jasmainak) [inline mne/preprocessing/maxwell.py]: same here. I don't see the argument

**Comment by jasmainak** [comment]
Other than this, I'm not able to give you much feedback @Eric89GXL . Thanks for the contrib

File: mne/preprocessing/maxwell.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/preprocessing/maxwell.py]: private functions are not testing by docstring tests

**Comment by Eric89GXL** [comment]
Thanks for the reviews. @wronk plans on looking soon and once he's happy I'll merge.

File: mne/preprocessing/maxwell.py [hunk 3]
```diff
@@ -678,6 +650,207 @@ def _sss_basis(origin, coils, int_order, ext_order, mag_scale=100.,
...
+    out_type = np.float64
+    n_coils = len(coils)
+
+    # Set up output matrices
+    n_in, n_out = _get_n_moments([int_order, ext_order])
+    S_tot = np.empty((len(coils), n_in + n_out), out_type)
+    S_in = S_tot[:, :n_in]
+    S_out = S_tot[:, n_in:]
+
+    # Set all magnetometers (with 'coil_class' == 1.0) to be scaled by 100
```
Comment (wronk) [inline mne/preprocessing/maxwell.py]: To fit it all in one line:
Scale all magnetometers (with `coil_class` == 1.0) by `mag_scale`

File: mne/preprocessing/tests/test_maxwell.py [hunk 4]
```diff
@@ -193,9 +192,15 @@ def test_multipolar_bases():
         assert_allclose(S_tot_mat, S_tot_mat_round, atol=1e-7)
         assert_allclose(S_tot_complex, S_tot_mat, rtol=1e-4, atol=1e-8)
 
+        # Now test our optimized version
+        S_tot = _sss_basis_basic(origin, coils, int_order, ext_order)
+        S_tot_fast = _sss_basis(origin, coils, int_order, ext_order)
+        # there are some sign differences in here...
```
Comment (wronk) [inline mne/preprocessing/tests/test_maxwell.py]: Is this a problem if they don't give the exact same results?

Comment (Eric89GXL) [inline mne/preprocessing/tests/test_maxwell.py]: I don't *think* it should be. The basis should still be a basis even if the signs are flipped. I suspect it's a Condon-Shortley phase thing.

Comment (Eric89GXL) [inline mne/preprocessing/tests/test_maxwell.py]: (the sign flip is uniform for a given column [or row, can't remember the orientation now])

Comment (wronk) [inline mne/preprocessing/tests/test_maxwell.py]: Cool, can you just note that in the comments?

**Comment by wronk** [comment]
LGTM once the above are added. Sorry this took me a bit. I was out of town for the holiday :)

**Comment by Eric89GXL** [comment]
Comments addressed @wronk, thanks. I'll merge once the CIs come back happy.

**Pull request merged**